[JCLOUDS-1377] add support for injectable Neutron Context into Nova

andreaturli opened the pull request:
cc @nacx

nacx commented on apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/config/NovaComputeServiceContextModule.java:
```diff
@@ -149,11 +161,34 @@ protected void configure() {
+      private final NovaSecurityGroupExtension novaSecurityGroupExtension;
+
+      @Inject
+      public SecurityGroupExtensionProvider(NeutronSecurityGroupExtension neutronSecurityGroupExtension,
```
Remove the public modifier

nacx commented on apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/extensions/NeutronSecurityGroupExtension.java:
```diff
@@ -0,0 +1,382 @@
+
+   @Inject(optional = true)
+   @Named("openstack-neutron")
+   Supplier<Context> neutronContextSupplier;
```
Make it `private`

nacx commented on apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/extensions/NeutronSecurityGroupExtension.java:
```diff
@@ -0,0 +1,382 @@
+   Supplier<Context> neutronContextSupplier;
+
+   @Inject
+   public NeutronSecurityGroupExtension(NovaApi api,
```
Remove the public modifier

nacx commented on apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/extensions/NeutronSecurityGroupExtension.java:
```diff
@@ -0,0 +1,382 @@
+      this.regionIds = checkNotNull(regionIds, "regionIds");
+      this.namingConvention = checkNotNull(namingConvention, "namingConvention");
+      this.groupCreator = groupCreator;
+      this.locationIndex = checkNotNull(locationIndex, "locationIndex");
```
Remove the null checks. They're redundant.

nacx commented on apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/extensions/NeutronSecurityGroupExtension.java:
```diff
@@ -0,0 +1,382 @@
+//      final FluentIterable<org.jclouds.openstack.nova.v2_0.domain.SecurityGroup> allGroups = securityGroupApi.listSecurityGroups();
+//      SecurityGroupInRegion rawGroup = new SecurityGroupInRegion(securityGroupApi.get(groupId), region, allGroups);
+
+//      return groupConverter.apply(rawGroup);
```
Remove the commented code.

nacx commented on apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/extensions/NeutronSecurityGroupExtension.java:
```diff
@@ -0,0 +1,382 @@
+      return false;
+   }
+
+   private Function<org.jclouds.openstack.neutron.v2.domain.SecurityGroup, SecurityGroup> toSecurityGroup(final Location location) {
```
Move this to its own type.
The private method is a bit awkward. Create a type for this function and use google assisted inject and a factory interface to populate the Location field where needed.

nacx commented on apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/functions/CleanupResources.java:
```diff
@@ -19,21 +19,27 @@
 import org.jclouds.logging.Logger;
 import org.jclouds.openstack.nova.v2_0.NovaApi;
-import org.jclouds.openstack.nova.v2_0.domain.SecurityGroup;
+import org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtension;
```
Unused?

nacx commented on apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/functions/CreateSecurityGroupIfNeeded.java:
```diff
@@ -17,73 +17,202 @@
+      return ((ApiContext<NeutronApi>) neutronContextSupplier.get()).getApi().getSecurityGroupApi(region);
+   }
+
+   // TODO remove duplications
```
You should be able to remove this when you have the function in a new type using assisted inject.

nacx commented on apis/openstack-nova/src/test/java/org/jclouds/openstack/nova/v2_0/compute/NovaComputeServiceLiveTest.java:
```diff
@@ -49,15 +62,22 @@ public void testOptionToNotBlock() {
+   @Override
+   public void testAScriptExecutionAfterBootWithBasicTemplate() throws Exception {
+      super.testAScriptExecutionAfterBootWithBasicTemplate();
+   }
```
Remove this.

nacx commented on apis/openstack-nova/src/test/java/org/jclouds/openstack/nova/v2_0/compute/extensions/NeutronSecurityGroupExtensionLiveTest.java:
```diff
@@ -0,0 +1,76 @@
+               "openstack-neutron.endpoint"))
+               .credentials(setIfTestSystemPropertyPresent(overrides,
+               "openstack-neutron.identity"),
+               setIfTestSystemPropertyPresent(overrides, "openstack-neutron.credential"))
```
We should be able to use the same Nova properties. It is unlikely to use a neutron that is behind a different keystone instance. Let's use the same endpoint and credentials that are configured for Nova.

nacx commented on apis/openstack-nova/src/test/java/org/jclouds/openstack/nova/v2_0/compute/extensions/NeutronSecurityGroupExtensionLiveTest.java:
```diff
@@ -0,0 +1,76 @@
+              .modules(ImmutableSet.<Module>of(
+                      new SshjSshClientModule(),
+                      new SLF4JLoggingModule(),
+                      new BouncyCastleCryptoModule())
```
Do we really need this module?

andreaturli commented:
@nacx thanks for the review

Main reason I have 2 couples of `Create/FindSecurityGroup` now (I want to remove them before merging this PR) is because of `NovaSecurityGroupExtension` and `NeutronSecurityGroupExtension`
`NovaSecurityGroupExtension` needs to use `NovaApi` to create security groups, while `NeutronSecurityGroupExtension` uses `NeutronApi`
Each couple of `Create/FindSecurityGroup` is a reasonably simple although ugly way to force each SecurityGroupExtension to use the appropriate API

Maybe a different way to solve the problem is to decide whether the SecurityGroupExtension needs to re-use the caching mechanism, wdyt?

andreaturli commented:
@nacx I think I've addressed most of your comments and now the duplication of `Find/Create` SecurityGroup is gone.

I still have some expected tests to fix but please let me know if the code is more readable now. Thanks

andreaturli commented:
rebuild please

nacx commented on apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/extensions/NeutronSecurityGroupExtension.java:
```diff
@@ -0,0 +1,334 @@
+      Optional<? extends ServerWithSecurityGroupsApi> serverApi = api.getServerWithSecurityGroupsApi(region);
+      SecurityGroupApi securityGroupApi = getSecurityGroupApi(region);
+
+      ServerWithSecurityGroups instance = serverApi.get().get(instanceId);
```
This is not used later. You're always returning all security groups.

andreaturli commented:
here's my live tests results:
- for `NovaSecurityGroupExtension`
```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=256m; support was removed in 8.0
Running org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest
Configuring TestNG with: TestNG652Configurator
Starting test testCreateSecurityGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest)
[TestNG] Test testCreateSecurityGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest) succeeded: 2618ms
Test suite progress: tests succeeded: 1, failed: 0, skipped: 0.
Starting test testSecurityGroupCacheInvalidated(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest)
[TestNG] Test testSecurityGroupCacheInvalidated(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest) succeeded: 5762ms
Test suite progress: tests succeeded: 2, failed: 0, skipped: 0.
Starting test testListSecurityGroups(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest)
INFO: Loading security groups
INFO: Found 4 security groups
INFO: Adding security group 1806862
INFO: Loading security groups
INFO: Found 5 security groups
INFO: Removing jclouds-1806862
INFO: Loading security groups
INFO: Found 4 security groups
[TestNG] Test testListSecurityGroups(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest) succeeded: 3267ms
Test suite progress: tests succeeded: 3, failed: 0, skipped: 0.
Starting test testSecurityGroupCacheInvalidatedWhenDeletedExternally(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest)
[TestNG] Test testSecurityGroupCacheInvalidatedWhenDeletedExternally(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest) succeeded: 6281ms
Test suite progress: tests succeeded: 4, failed: 0, skipped: 0.
Starting test testGetSecurityGroupById(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest)
[TestNG] Test testGetSecurityGroupById(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest) succeeded: 474ms
Test suite progress: tests succeeded: 5, failed: 0, skipped: 0.
Starting test testAddIpPermission(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest)
[TestNG] Test testAddIpPermission(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest) succeeded: 2004ms
Test suite progress: tests succeeded: 6, failed: 0, skipped: 0.
Starting test testRemoveIpPermission(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest)
[TestNG] Test testRemoveIpPermission(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest) succeeded: 2080ms
Test suite progress: tests succeeded: 7, failed: 0, skipped: 0.
Starting test testAddIpPermissionsFromSpec(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest)
[TestNG] Test testAddIpPermissionsFromSpec(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest) succeeded: 2295ms
Test suite progress: tests succeeded: 8, failed: 0, skipped: 0.
Starting test testAddIpPermissionWithCidrExclusionGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest)
[TestNG] Test testAddIpPermissionWithCidrExclusionGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest) skipped.
Test suite progress: tests succeeded: 8, failed: 0, skipped: 1.
[TestNG] Test testRemoveIpPermissionWithCidrExclusionGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest) skipped.
Test suite progress: tests succeeded: 8, failed: 0, skipped: 2.
Starting test testDeleteSecurityGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest)
[TestNG] Test testDeleteSecurityGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest) succeeded: 718ms
Test suite progress: tests succeeded: 9, failed: 0, skipped: 2.
Tests run: 11, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 27.195 sec - in org.jclouds.openstack.nova.v2_0.compute.extensions.NovaSecurityGroupExtensionLiveTest

Results :

Tests run: 11, Failures: 0, Errors: 0, Skipped: 2
```
and for `NeutronSecurityGroupExtensions` with Neutron-context attached
```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest
Configuring TestNG with: TestNG652Configurator
Starting test testSecurityGroupCacheInvalidated(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
Starting test testCreateSecurityGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
Starting test testSecurityGroupCacheInvalidatedWhenDeletedExternally(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
[pool-2-thread-1] Test testCreateSecurityGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) succeeded: 10400ms
Test suite progress: tests succeeded: 1, failed: 0, skipped: 0.
Starting test testGetSecurityGroupById(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
[pool-2-thread-4] Test testGetSecurityGroupById(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) succeeded: 584ms
Test suite progress: tests succeeded: 2, failed: 0, skipped: 0.
Starting test testAddIpPermission(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
[pool-2-thread-3] Test testSecurityGroupCacheInvalidatedWhenDeletedExternally(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) failed.
Test suite progress: tests succeeded: 2, failed: 1, skipped: 0.
[pool-2-thread-5] Test testAddIpPermission(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) succeeded: 2758ms
Test suite progress: tests succeeded: 3, failed: 1, skipped: 0.
Starting test testRemoveIpPermission(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
[pool-2-thread-2] Test testSecurityGroupCacheInvalidated(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) succeeded: 17480ms
Test suite progress: tests succeeded: 4, failed: 1, skipped: 0.
[pool-2-thread-5] Test testRemoveIpPermission(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) succeeded: 4162ms
Test suite progress: tests succeeded: 5, failed: 1, skipped: 0.
Starting test testAddIpPermissionsFromSpec(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
[pool-2-thread-5] Test testAddIpPermissionsFromSpec(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) succeeded: 2849ms
Test suite progress: tests succeeded: 6, failed: 1, skipped: 0.
Starting test testAddIpPermissionWithCidrExclusionGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
[pool-2-thread-5] Test testAddIpPermissionWithCidrExclusionGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) skipped.
Test suite progress: tests succeeded: 6, failed: 1, skipped: 1.
[pool-2-thread-5] Test testRemoveIpPermissionWithCidrExclusionGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) skipped.
Test suite progress: tests succeeded: 6, failed: 1, skipped: 2.
Starting test testDeleteSecurityGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
[pool-2-thread-5] Test testDeleteSecurityGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) succeeded: 1334ms
Test suite progress: tests succeeded: 7, failed: 1, skipped: 2.
Tests run: 10, Failures: 1, Errors: 0, Skipped: 2, Time elapsed: 24.094 sec <<< FAILURE! - in org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest
testSecurityGroupCacheInvalidatedWhenDeletedExternally(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)  Time elapsed: 2.308 sec  <<< FAILURE!
java.lang.IllegalStateException: command: DELETE https://networking.uk-1.cloud.global.fujitsu.com/v2.0/security-groups/341254d6-4891-4afc-ad68-20aac0e881d4 HTTP/1.1 failed with response: HTTP/1.1 409 Conflict; content: [{"NeutronError": {"message": "Operation conflicts with other operation", "type": "Conflict", "detail": ""}}]
	at org.jclouds.openstack.neutron.v2.handlers.NeutronErrorHandler.handleError(NeutronErrorHandler.java:57)
	at org.jclouds.http.handlers.DelegatingErrorHandler.handleError(DelegatingErrorHandler.java:65)
	at org.jclouds.http.internal.BaseHttpCommandExecutorService.shouldContinue(BaseHttpCommandExecutorService.java:138)
	at org.jclouds.http.internal.BaseHttpCommandExecutorService.invoke(BaseHttpCommandExecutorService.java:107)
	at org.jclouds.rest.internal.InvokeHttpMethod.invoke(InvokeHttpMethod.java:91)
	at org.jclouds.rest.internal.InvokeHttpMethod.apply(InvokeHttpMethod.java:74)
	at org.jclouds.rest.internal.InvokeHttpMethod.apply(InvokeHttpMethod.java:45)
	at org.jclouds.reflect.FunctionalReflection$FunctionalInvocationHandler.handleInvocation(FunctionalReflection.java:117)
	at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:87)
	at com.sun.proxy.$Proxy86.deleteSecurityGroup(Unknown Source)
	at org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtension.removeSecurityGroup(NeutronSecurityGroupExtension.java:168)
	at org.jclouds.compute.extensions.internal.BaseSecurityGroupExtensionLiveTest.deleteSecurityGroupFromAnotherView(BaseSecurityGroupExtensionLiveTest.java:441)
	at org.jclouds.compute.extensions.internal.BaseSecurityGroupExtensionLiveTest.testSecurityGroupCacheInvalidatedWhenDeletedExternally(BaseSecurityGroupExtensionLiveTest.java:415)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.jclouds.http.HttpResponseException: command: DELETE https://networking.uk-1.cloud.global.fujitsu.com/v2.0/security-groups/341254d6-4891-4afc-ad68-20aac0e881d4 HTTP/1.1 failed with response: HTTP/1.1 409 Conflict; content: [{"NeutronError": {"message": "Operation conflicts with other operation", "type": "Conflict", "detail": ""}}]
	at org.jclouds.openstack.neutron.v2.handlers.NeutronErrorHandler.handleError(NeutronErrorHandler.java:40)
	at org.jclouds.http.handlers.DelegatingErrorHandler.handleError(DelegatingErrorHandler.java:65)
	at org.jclouds.http.internal.BaseHttpCommandExecutorService.shouldContinue(BaseHttpCommandExecutorService.java:138)
	at org.jclouds.http.internal.BaseHttpCommandExecutorService.invoke(BaseHttpCommandExecutorService.java:107)
	at org.jclouds.rest.internal.InvokeHttpMethod.invoke(InvokeHttpMethod.java:91)
	at org.jclouds.rest.internal.InvokeHttpMethod.apply(InvokeHttpMethod.java:74)
	at org.jclouds.rest.internal.InvokeHttpMethod.apply(InvokeHttpMethod.java:45)
	at org.jclouds.reflect.FunctionalReflection$FunctionalInvocationHandler.handleInvocation(FunctionalReflection.java:117)
	at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:87)
	at com.sun.proxy.$Proxy86.deleteSecurityGroup(Unknown Source)
	at org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtension.removeSecurityGroup(NeutronSecurityGroupExtension.java:168)
	at org.jclouds.compute.extensions.internal.BaseSecurityGroupExtensionLiveTest.deleteSecurityGroupFromAnotherView(BaseSecurityGroupExtensionLiveTest.java:441)
	at org.jclouds.compute.extensions.internal.BaseSecurityGroupExtensionLiveTest.testSecurityGroupCacheInvalidatedWhenDeletedExternally(BaseSecurityGroupExtensionLiveTest.java:415)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)


Results :

Failed tests:
  NeutronSecurityGroupExtensionLiveTest>BaseSecurityGroupExtensionLiveTest.testSecurityGroupCacheInvalidatedWhenDeletedExternally:415->BaseSecurityGroupExtensionLiveTest.deleteSecurityGroupFromAnotherView:441 » IllegalState

Tests run: 10, Failures: 1, Errors: 0, Skipped: 2
```

I need to understand why there is that failure

nacx commented on apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/extensions/NeutronSecurityGroupExtension.java:
```diff
@@ -0,0 +1,325 @@
+      final FluentIterable<org.jclouds.openstack.neutron.v2.domain.SecurityGroup> allGroups = securityGroupApi.listSecurityGroups().concat();
+      Location location = locationIndex.get().get(region);
+      return ImmutableSet.copyOf(transform(filter(allGroups, notNull()), neutronSecurityGroupToSecurityGroup.create(location)));
+   }
```
This implementation is still wrong. You have to filter ONLY the security groups associated with the given node. You're returning all existing ones.

andreaturli replied:
good point! thanks

unfortunately also `NovaSecurityGroupExtension` is not really reliable as it is built on
`Optional<? extends ServerWithSecurityGroupsApi> serverWithSecurityGroupsApi = api.getServerWithSecurityGroupsApi(region)` which is most of the time not working on recent Openstack installations (they claim the extension is there but it is not working!)

We may need to radically change the implementation but not sure how yet

andreaturli replied:
I think we should use https://developer.openstack.org/api-ref/compute/#servers-security-groups-servers-os-security-groups

nacx replied:
+1 to use this. I'd say we need it before merging the PR. Agree?

andreaturli commented:
@nacx here's the result of `NeutronSecurityGroupExtensionLiveTest` against a provider with keystone v3 and Neutron support
```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest
Configuring TestNG with: TestNG652Configurator
Starting test testCreateSecurityGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
Starting test testSecurityGroupCacheInvalidatedWhenDeletedExternally(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
Starting test testSecurityGroupCacheInvalidated(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
[pool-2-thread-1] Test testCreateSecurityGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) succeeded: 10013ms
Test suite progress: tests succeeded: 1, failed: 0, skipped: 0.
Starting test testGetSecurityGroupById(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
[pool-2-thread-4] Test testGetSecurityGroupById(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) succeeded: 560ms
Test suite progress: tests succeeded: 2, failed: 0, skipped: 0.
Starting test testAddIpPermission(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
[pool-2-thread-5] Test testAddIpPermission(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) succeeded: 2908ms
Test suite progress: tests succeeded: 3, failed: 0, skipped: 0.
Starting test testRemoveIpPermission(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
[pool-2-thread-3] Test testSecurityGroupCacheInvalidatedWhenDeletedExternally(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) succeeded: 16488ms
Test suite progress: tests succeeded: 4, failed: 0, skipped: 0.
[pool-2-thread-2] Test testSecurityGroupCacheInvalidated(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) succeeded: 16835ms
Test suite progress: tests succeeded: 5, failed: 0, skipped: 0.
[pool-2-thread-5] Test testRemoveIpPermission(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) succeeded: 3464ms
Test suite progress: tests succeeded: 6, failed: 0, skipped: 0.
Starting test testAddIpPermissionsFromSpec(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
[pool-2-thread-5] Test testAddIpPermissionsFromSpec(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) succeeded: 2700ms
Test suite progress: tests succeeded: 7, failed: 0, skipped: 0.
Starting test testAddIpPermissionWithCidrExclusionGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
[pool-2-thread-5] Test testAddIpPermissionWithCidrExclusionGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) skipped.
Test suite progress: tests succeeded: 7, failed: 0, skipped: 1.
[pool-2-thread-5] Test testRemoveIpPermissionWithCidrExclusionGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) skipped.
Test suite progress: tests succeeded: 7, failed: 0, skipped: 2.
Starting test testDeleteSecurityGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest)
[pool-2-thread-5] Test testDeleteSecurityGroup(org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest) succeeded: 1202ms
Test suite progress: tests succeeded: 8, failed: 0, skipped: 2.
Tests run: 10, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 22.939 sec - in org.jclouds.openstack.nova.v2_0.compute.extensions.NeutronSecurityGroupExtensionLiveTest

Results :

Tests run: 10, Failures: 0, Errors: 0, Skipped: 2
```

nacx commented on apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/config/NovaComputeServiceContextModule.java:
```diff
@@ -203,14 +232,14 @@ protected TemplateOptions provideTemplateOptions(Injector injector, TemplateOpti
+//            FindSecurityGroupWithNameAndReturnTrue in,
+//            @Named(TIMEOUT_SECURITYGROUP_PRESENT) long msDelay) {
+//      return retry(in, msDelay, 100L, MILLISECONDS);
+//   }
```
Remove the commented code.

nacx commented on apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/functions/NovaSecurityGroupToSecurityGroup.java:
```diff
@@ -0,0 +1,130 @@
+       * builder.cidrBlock(from.getRemoteIpPrefix()); }
+       * 
+       * return builder.build(); } }), Predicates.notNull())); }
+       */
```
Remove the commented code.

nacx commented on apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/functions/NovaSecurityGroupToSecurityGroup.java:
```diff
@@ -0,0 +1,130 @@
+       * , ruleGroup, owningGroup.getName(), owningGroup.getId()); return null; }
+       * builder.groupId(group.getRegion() + "/" +
+       * Iterables.getOnlyElement(referredGroup).getId()); }
+       */
```
Remove the commented code.

nacx commented on apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/strategy/ApplyNovaTemplateOptionsCreateNodesWithGroupEncodedIntoNameThenAddToSet.java:
```diff
@@ -132,29 +129,32 @@ protected ApplyNovaTemplateOptionsCreateNodesWithGroupEncodedIntoNameThenAddToSe
-         templateOptions.securityGroups(securityGroupName);
-         tagsBuilder.add(String.format("%s-%s", JCLOUDS_SG_PREFIX, securityGroupInRegion.getSecurityGroup().getId()));
+
+         if (securityGroup == null) {
```
At this point this is always null. Remove the dedundant `if` and declare the variable here.

andreaturli commented:
merging it

andreaturli commented:
thanks @nacx for your help and review

andreaturli commented:
merged at [master](https://git-wip-us.apache.org/repos/asf?p=jclouds.git;h=09936b5)
